Implement human-readable serde for `Witness`

tcharding opened the pull request:
This is dr-orlovsky's [PR](https://github.com/rust-bitcoin/rust-bitcoin/pull/899) picked up at his permission in the discussion thread.

I went through the review comments and implemented everything except the perf optimisations. Also includes a patch at the front of the PR that adds a unit test that can be run to see the "before and after", not sure if we want it in, perhaps it should be removed before merge.

This PR implicitly fixes 942.

To test this PR works as advertised run `cargo test display_transaction --features=serde -- --nocapture` after creating a unit test as follows: 
```rust

    // Used to verify that parts of a transaction pretty print.
    // `cargo test display_transaction --features=serde -- --nocapture`
    #[cfg(feature = "serde")]
    #[test]
    fn serde_display_transaction() {
        let tx_bytes = Vec::from_hex(
            "02000000000101595895ea20179de87052b4046dfe6fd515860505d6511a9004cf12a1f93cac7c01000000\
            00ffffffff01deb807000000000017a9140f3444e271620c736808aa7b33e370bd87cb5a078702483045022\
            100fb60dad8df4af2841adc0346638c16d0b8035f5e3f3753b88db122e70c79f9370220756e6633b17fd271\
            0e626347d28d60b0a2d6cbb41de51740644b9fb3ba7751040121028fa937ca8cba2197a37c007176ed89410\
            55d3bcb8627d085e94553e62f057dcc00000000"
        ).unwrap();
        let tx: Transaction = deserialize(&tx_bytes).unwrap();
        let ser = serde_json::to_string_pretty(&tx).unwrap();
        println!("{}", ser);
    }
```

Fixes: #942

RCasatta commented on src/blockdata/witness.rs:
```diff
@@ -282,18 +281,69 @@ impl serde::Serialize for Witness {
     {
-        let vec: Vec<_> = self.to_vec();
-        serde::Serialize::serialize(&vec, serializer)
+        use hashes::hex::ToHex;
+        use serde::ser::SerializeSeq;
+
+        if serializer.is_human_readable() {
+            let mut seq = serializer.serialize_seq(Some(self.witness_elements))?;
+            for elem in self.iter() {
+                seq.serialize_element(&elem.to_hex())?;
+            }
+            seq.end()
+        } else {
+            let vec: Vec<_> = self.to_vec();
+            serde::Serialize::serialize(&vec, serializer)
+        }
```
We should do the same `serialize_element` loop for non-human-readable to avoid `to_vec()`

```rust
let human_readable = serializer.is_human_readable();
let mut seq = serializer.serialize_seq(Some(self.witness_elements))?;

for elem in self.iter() {
    if human_readable {
        seq.serialize_element(&elem.to_hex())?;
    } else {
        seq.serialize_element(&elem)?;
    }
}
seq.end()
```

DanGould replied:
Yes please, this solves https://github.com/rust-bitcoin/rust-bitcoin/issues/942

tcharding replied:
Good idea, will do. Thanks

DanGould commented on src/blockdata/witness.rs:
```diff
@@ -282,18 +281,69 @@ impl serde::Serialize for Witness {
-        let vec: Vec<_> = self.to_vec();
-        serde::Serialize::serialize(&vec, serializer)
+        use hashes::hex::ToHex;
+        use serde::ser::SerializeSeq;
```
I'm curious, why are these `use` statements put here rather than the top of the file?

tcharding replied:
woops, I had a stale view of this page when I wrote my answer, I did not see @dr-orlovsky's answer at that time.

tcharding commented:
Changes in force push:
- Use loop suggestion from @RCasatta to remove allocation

Kixunil reviewed (commented):
Quick review for now.

tcharding commented on src/blockdata/witness.rs:
```diff
@@ -282,18 +279,70 @@ impl serde::Serialize for Witness {
+                    let vec = Vec::<u8>::from_hex(&elem).map_err(|e| {
+                        match e {
+                            InvalidChar(c) => {
+                                let c = core::char::from_u32(c as u32).unwrap_or('-');
```
Sure thing, I think you told me this before. I must not have fully groked it then, so I went looking :) Is it because `as` can be lossy and `from` only does lossless conversions?

ref: https://stackoverflow.com/questions/48795329/what-is-the-difference-between-fromfrom-and-as-in-rust

Kixunil replied:
Why this still contains `as` and was marked resolved? Also still uses `-`.

tcharding replied:
Seems that fresh mountain air on my week off broke my brain, I thought we had a discussion about no using default '-' and I don't see that. Not to mention that the '-' is still there like you say. Anyways, I've removed it and use `Unexpected::Unsigned` when its not a char (like I remember you suggesting previously). Thanks for patiently, repeatedly pointing out my mistakes :)

tcharding commented on src/blockdata/witness.rs:
```diff
@@ -282,18 +279,70 @@ impl serde::Serialize for Witness {
+                        match e {
+                            InvalidChar(c) => {
+                                let c = core::char::from_u32(c as u32).unwrap_or('-');
+                                serde::de::Error::invalid_value(serde::de::Unexpected::Char(c), &"a valid hex character")
```
Nice, thanks man, I didn't really like the `'-'`.

tcharding commented on src/blockdata/witness.rs:
```diff
@@ -282,18 +279,70 @@ impl serde::Serialize for Witness {
+        if deserializer.is_human_readable() {
+            deserializer.deserialize_seq(Visitor)
+        } else {
+            let vec: Vec<Vec<u8>> = serde::Deserialize::deserialize(deserializer)?;
```
Righto, I'll have a play with it.

tcharding replied:
I must not be fully understanding where all the allocations are occurring. I came up with the code below but it seems to me that this code is going to contain the exact same number of allocations as line 343 above (one for each element, and the one outer vector).
```rust
            fn visit_seq<A: serde::de::SeqAccess<'de>>(self, mut a: A) -> Result<Self::Value, A::Error>
            {
                let mut ret = Vec::new();
                while let Some(elem) = a.next_element::<Vec<u8>>()? {
                    ret.push(elem);
                }
                Ok(Witness::from_vec(ret))
            }
```
What am I missing?

Kixunil replied:
I meant to build `Witness` as non-nested `Vec` directly. You need to use `next_element_seed` and a bit more machinery to support it. [This example](https://docs.rs/serde/latest/serde/de/trait.DeserializeSeed.html#example) looks *very* similar to what we have here, you can pretty much copy it minus some generics (we have `u8`). Note that that example forgot to call `reserve()` in front of the inner `for` loop. Should have this:

```rust
if let Some(size_hint) = seq.size_hint() {
    self.0.reserve(size_hint);
}
```

tcharding replied:
I had a good play with this, was educational. Its separate from this PR because its an optimisation, right? I think the correct way to go about it is merge this naive version (fixes the ugly serialization) and then create an separate issue/PR to do it as in the example you link. Would that be fair?

Kixunil replied:
I don't care if it's merged in this one or a separate one.

tcharding replied:
Cool, I'll do it as a separate PR. Cheers.

tcharding commented on src/blockdata/witness.rs:
```diff
@@ -282,18 +279,70 @@ impl serde::Serialize for Witness {
+                            OddLengthString(len) => serde::de::Error::invalid_length(len, &"an even length string"),
+                            InvalidLength(expected, got) => {
+                                let exp = format!("expected length: {}", expected);
+                                serde::de::Error::invalid_length(got, &exp.as_str())
```
:)

tcharding replied:
Lolz, everything builds when one doesn't enable the feature. This `as_str` is needed, without it we get:
```
error[E0277]: the trait bound `std::string::String: serde::de::Visitor<'_>` is not satisfied
   --> src/blockdata/witness.rs:335:71
    |
335 | ...                   serde::de::Error::invalid_length(got, &exp)
    |                       --------------------------------      ^^^^ the trait `serde::de::Visitor<'_>` is not implemented for `std::string::String`
    |                       |
    |                       required by a bound introduced by this call
```

Kixunil replied:
Oh, I wrongly remembered the second argument is valid for all `T: Display`, sorry about the noise.

(It's the other way around, `&dyn Expected` impls `Display` and `Expected` is available for `T: Visitor + Sized` and `str`)

apoelstra commented:
I suspect that serializaiton could also avoid the `to_hex` allocation -- but yes, let's go ahead and do that in another PR :)

apoelstra reviewed (approved):
ACK 4c1dfafcdd8284685fbb0bf880d3cb47bb46d541

Kixunil commented:
@apoelstra It's not a clear win - if the serializer uses the default implementation we end up having *slower* code. I just [posted to IRLO about this](https://internals.rust-lang.org/t/size-hint-for-display-and-other-fmt-traits/17004)

Kixunil commented on src/blockdata/transaction.rs:
```diff
@@ -1733,6 +1733,23 @@ mod tests {
+        ).unwrap();
+        let tx: Transaction = deserialize(&tx_bytes).unwrap();
+        let ser = serde_json::to_string_pretty(&tx).unwrap();
+        println!("{}", ser);
```
I think we should use `assert_eq`. I know changes to serde would break this but seems less bad than manually checking with funny flags.

tcharding replied:
I decided to remove the test all together. I put it in the PR description in case anyone wants to verify that this PR works as advertised.

Kixunil commented on src/blockdata/witness.rs:
```diff
@@ -279,25 +279,70 @@ impl serde::Serialize for Witness {
+                use hashes::hex::FromHex;
+                use hashes::hex::Error::*;
+
+                let mut ret = Vec::new();
```
`with_capacity`

tcharding replied:
Done, cheers.

Kixunil reviewed (commented):
Were some changes lost by accident?

tcharding commented:
Changes in force push:
- Dropped patch that added the unit test displaying `Witness`
- Modified commit `68577dfb Witness human-readable serde` by doing
  - Added import for serde stuff to shorten types/lines
  - Removed use of `unwrap_or('-')`, used multiple different serde error types
  - Used `into` instead of `as`

Kixunil reviewed (approved):
ACK a1df62a3d927f0a8b66228df5c3db422e220ac03
